x/accounts/internal/implementation/extension.go

```diff
@@ -0,0 +1,74 @@
+package implementation
+
+import (
+	"context"
+	"fmt"
+	"iter"
+	"maps"
+
+	gogoproto "github.com/cosmos/gogoproto/proto"
+
+	"cosmossdk.io/core/transaction"
+)
+
+var _ ProtoMsgHandlerRegistry = &ExtensionExecuteAdapter{}
+
+type ProtoMsgHandler func(ctx context.Context, msg transaction.Msg) (resp transaction.Msg, err error)
+
+type ExtensionExecuteAdapter struct {
+	// handlers is a map of handler functions that will be called when the smart account is executed.
+	handlers map[string]ProtoMsgHandler
+
+	// protoSchemas is a map of schemas for the messages that will be passed to the handler functions
+	// and the messages that will be returned by the handler functions.
+	protoSchemas map[string]HandlerSchema
+}
+
+func NewExtensionExecuteAdapter() *ExtensionExecuteAdapter {
+	return &ExtensionExecuteAdapter{
+		handlers:     make(map[string]ProtoMsgHandler),
+		protoSchemas: make(map[string]HandlerSchema),
+	}
+}
+
+func (e ExtensionExecuteAdapter) ExtendHandlers(accountExec ProtoMsgHandler) (ProtoMsgHandler, error) {
+	return func(ctx context.Context, msg transaction.Msg) (resp transaction.Msg, err error) {
+		if len(e.handlers) != 0 {
+			messageName := MessageName(msg)
+			handler, ok := e.handlers[messageName]
+			if ok {
```

Comment:
when a matching extension handler exists, it is used. Otherwise, fall though to the account handlers